### PR TITLE
[feature](compaction) support delete in cumulative compaction

### PR DESCRIPTION
**Pull request by ByteYue**
# Proposed changes

Issue Number: close #xxx
Before this pr, cumulative compaction can not do delete, which means the cumu compaction policy would pick input rowset which has no delete predicate and let the rowset with delete predication to be processed by base compaction. Now the delete predication is carried within the rowset meta, so actually it's feasible for us to merge the delete predications in multiple input rowsets into the output rowsets. Once we make the delete handler in cumu compaction initialized, the remaining data's delete procedure is the same with base compaction.
Future work: for merge on write, maybe we need add some rows num check logic.
https://selectdb.feishu.cn/docx/TIxjdXILAoO51fxXlvQc06hknAP
## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by ByteYue** [comment]
run beut

**Comment by ByteYue** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 33.6 seconds
 stream load tsv:          430 seconds loaded 74807831229 Bytes, about 165 MB/s
 stream load json:         24 seconds loaded 2358488459 Bytes, about 93 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230513085830_clickbench_pr_143279.html

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by gavinchou** [comment]
LGTM

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 43.05 seconds
 stream load tsv:          449 seconds loaded 74807831229 Bytes, about 158 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          58 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          28 seconds loaded 861443392 Bytes, about 29 MB/s
 insert into select:          69.4 seconds inserted 10000000 Rows, about 144K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230630135021_clickbench_pr_170583.html

File: be/src/common/config.cpp [hunk 1]
```diff
@@ -1023,6 +1023,8 @@ DEFINE_Bool(enable_set_in_bitmap_value, "false");
 
 DEFINE_Int64(max_hdfs_file_handle_cache_num, "20000");
 DEFINE_Int64(max_external_file_meta_cache_num, "20000");
+// enable do delete in cumu compaction
```
Comment (platoneko) [inline be/src/common/config.cpp:1026]:
```suggestion
// Apply delete pred in cumu compaction
```

File: be/src/olap/compaction.cpp [hunk 2]
```diff
@@ -339,6 +340,28 @@ Status Compaction::do_compaction_impl(int64_t permits) {
                      << ", output_version=" << _output_version;
         return Status::Error<ROWSET_BUILDER_INIT>();
     }
+    // now we support delete in cumu compaction, to make all data in rowsets whose version
```
Comment (platoneko) [inline be/src/olap/compaction.cpp:343]: Capitalizes the first letter
```suggestion
    // Now ...
```

**Comment by ByteYue** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new mechine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 51.33 seconds
 stream load tsv:          509 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          89.5 seconds inserted 10000000 Rows, about 111K ops/s
 storage size: 17164533518 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230705231349_clickbench_pr_172963.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 48.76 seconds
 stream load tsv:          456 seconds loaded 74807831229 Bytes, about 156 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          58 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          29 seconds loaded 861443392 Bytes, about 28 MB/s
 insert into select:          69.2 seconds inserted 10000000 Rows, about 144K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230705153035_clickbench_pr_172959.html

**Review by platoneko — CHANGES_REQUESTED** [review]
```c++
Status BaseCompaction::pick_rowsets_to_compact() {
    ...
    if (!config::enable_delete_when_cumu_compaction && _input_rowsets.front()->start_version() > 0) {
        ...
```

**Comment by ByteYue** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 51.72 seconds
 stream load tsv:          508 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          87.0 seconds inserted 10000000 Rows, about 114K ops/s
 storage size: 17162898244 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230707120349_clickbench_pr_174120.html

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 51.2 seconds
 stream load tsv:          499 seconds loaded 74807831229 Bytes, about 142 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          87.5 seconds inserted 10000000 Rows, about 114K ops/s
 storage size: 17164075209 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230707121547_clickbench_pr_174172.html

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 57.54 seconds
 stream load tsv:          497 seconds loaded 74807831229 Bytes, about 143 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          89.2 seconds inserted 10000000 Rows, about 112K ops/s
 storage size: 17167662978 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230707121554_clickbench_pr_174156.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.95 seconds
 stream load tsv:          447 seconds loaded 74807831229 Bytes, about 159 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          56 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          69.4 seconds inserted 10000000 Rows, about 144K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230707054207_clickbench_pr_174169.html

**Review by platoneko — APPROVED** [review]

**Comment by ByteYue** [comment]
run beut

**Comment by ByteYue** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 56.99 seconds
 stream load tsv:          512 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17169245746 Bytes

**Comment by ByteYue** [comment]
run p0

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 50.65 seconds
 stream load tsv:          509 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          28.7 seconds inserted 10000000 Rows, about 348K ops/s
 storage size: 17162117574 Bytes

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 49.66 seconds
 stream load tsv:          537 seconds loaded 74807831229 Bytes, about 132 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          28.5 seconds inserted 10000000 Rows, about 350K ops/s
 storage size: 17166479556 Bytes

**Review by dataroaring — DISMISSED** [review]
LGTM

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 44.12 seconds
 stream load tsv:          506 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          33 seconds loaded 861443392 Bytes, about 24 MB/s
 insert into select:          29.4 seconds inserted 10000000 Rows, about 340K ops/s
 storage size: 17163999123 Bytes

**Comment by ByteYue** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.98 seconds
 stream load tsv:          513 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          29.4 seconds inserted 10000000 Rows, about 340K ops/s
 storage size: 17162512644 Bytes

**Review by platoneko — APPROVED** [review]

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**